File: tests/core/test_paths.py
from pathlib import Path

import pytest

from core.errors import ToolError
from core.paths import Workspace


def test_resolve_inside(workspace: Workspace) -> None:
    assert workspace.resolve("a/b.mp4") == workspace.root / "a" / "b.mp4"


@pytest.mark.parametrize("bad", ["../fora.mp4", "/etc/passwd", "a/../../x"])
def test_resolve_rejects_escape(workspace: Workspace, bad: str) -> None:
    with pytest.raises(ToolError) as exc:
        workspace.resolve(bad)
    assert exc.value.code == "outside_workspace"


def test_existing_missing(workspace: Workspace) -> None:
    with pytest.raises(ToolError) as exc:
        workspace.existing("nao.mp4")
    assert exc.value.code == "not_found"


def test_output_for_is_unique(workspace: Workspace) -> None:
    src = workspace.root / "v.mp4"
    src.touch()
    first = workspace.output_for(src, "cut 0-10")
    assert first.name == "v_cut_0-10.mp4"
    first.touch()
    second = workspace.output_for(src, "cut 0-10")
    assert second.name == "v_cut_0-10_1.mp4"


def test_output_for_extension(workspace: Workspace) -> None:
    src = Path(workspace.root / "v.mp4")
    assert workspace.output_for(src, "frame", extension="png").suffix == ".png"

File: version-in-rust/tests/core/test_paths.rs
use mcp_tools::core::errors::ErrorCode;

use crate::conftest::workspace;

#[test]
fn test_resolve_inside() {
    let ws = workspace();
    let resolved = ws.workspace.resolve("a/b.mp4").unwrap();
    assert_eq!(resolved, ws.workspace.root.join("a").join("b.mp4"));
}

#[test]
fn test_resolve_rejects_escape() {
    let ws = workspace();
    for bad in ["../fora.mp4", "/etc/passwd", "a/../../x"] {
        let error = ws.workspace.resolve(bad).unwrap_err();
        assert_eq!(error.code, ErrorCode::OutsideWorkspace, "{bad}");
    }
}

#[test]
fn test_existing_missing() {
    let ws = workspace();
    let error = ws.workspace.existing("nao.mp4").unwrap_err();
    assert_eq!(error.code, ErrorCode::NotFound);
}

#[test]
fn test_output_for_is_unique() {
    let ws = workspace();
    let src = ws.workspace.root.join("v.mp4");
    std::fs::write(&src, b"").unwrap();
    let first = ws.workspace.output_for(&src, "cut 0-10", None);
    assert_eq!(first.file_name().unwrap(), "v_cut_0-10.mp4");
    std::fs::write(&first, b"").unwrap();
    let second = ws.workspace.output_for(&src, "cut 0-10", None);
    assert_eq!(second.file_name().unwrap(), "v_cut_0-10_1.mp4");
}

#[test]
fn test_output_for_extension() {
    let ws = workspace();
    let src = ws.workspace.root.join("v.mp4");
    let out = ws.workspace.output_for(&src, "frame", Some("png"));
    assert_eq!(out.extension().unwrap(), "png");
}
